core/common/src/main/java/alluxio/underfs/UnderFileSystem.java

```diff
@@ -411,11 +411,30 @@ default AlluxioConfiguration getConfiguration() throws IOException {
    * file is only renamed (identical content and permissions). Returns
    * {@link alluxio.Constants#INVALID_UFS_FINGERPRINT} if there is any error.
    *
+   * @deprecated
    * @param path the path to compute the fingerprint for
    * @return the string representing the fingerprint
    */
+  @Deprecated
   String getFingerprint(String path);
 
+  /**
+   * Computes and returns a fingerprint for the path. The fingerprint is used to determine if two
+   * UFS files are identical. The fingerprint must be deterministic, and must not change if a
+   * file is only renamed (identical content and permissions). Returns
+   * {@link Fingerprint#INVALID_FINGERPRINT} if there is any error.
+   *
+   * The default implementation relies on {@link #getFingerprint(String)} and there is one extra
+   * parsing. This default implementation is mainly for backward compatibility.
+   * Override this for performance.
+   *
+   * @param path the path to compute the fingerprint for
+   * @return the string representing the fingerprint
+   */
+  default Fingerprint getParsedFingerprint(String path) {
```

Comment:
This is made `default` so existing children are not forced to change, and stay with the inefficient way. But the old way should be deprecated and removed. I'm not sure if there are better ways than just mentioning in the doc.